### DISCUSSION TITLE
Understanding Campbell Plot

### DISCUSSION THREAD

**Discussion by FadliM21** [petrobras/ross#1041] [Q&A]
Hi,

I'm a beginner for Rotor dynamics. Can you please help me to understand the campbell plot? What is the meaning of Forward, Mixed and Backward in campbell legend?

![campbell](https://github.com/petrobras/ross/assets/165719533/2ae55b71-3ee4-4066-a23f-e2c11747610c)

Thank you

**Comment by jguarato** [comment]
Hi, @FadliM21!

The Campbell diagram is a key tool in rotordynamics used to analyze the vibration characteristics of rotating machinery. The curves on the plot represent the natural frequencies (Y-axis) of the system as a function of rotor speed (X-axis). These frequencies change as the rotor speed increases due to the gyroscopic effect and other dynamic interactions within the system.

The terms "Forward", "Backward", and "Mixed" refer to the direction of rotation of the vibrational modes relative to the rotation of the rotor. The **Forward** mode occurs when the mode rotates in the same direction as the rotor spin. The **Backward** mode occurs when the mode rotates in the opposite direction to the rotor spin. The **Mixed** mode can occur when the vibration mode rotates in both forward and backward directions. Lastly, the **Critical Speeds** are the rotor speeds at which the natural frequencies coincide with the excitation frequencies (typically harmonics of the rotor speed).

For a more comprehensive understanding, we highly recommend consulting the reference book, specifically sections 3.5–3.6 and 5.9.

Apologies for the delayed response.

### **Reference:**

[🔗](https://doi.org/10.1017/CBO9780511780509.004) M. I. Friswell, J. E. T. Penny, S. D. Garvey, and A. W. Lees, *Dynamics of Rotating Machines*, Cambridge: Cambridge University Press, 2010, pp. 76–123.

**Reply by FadliM21** [reply]
Hi @jguarato 

May I ask another problem?
I've tried to simulate with real case our machine here, but I still have different result with the Vendor's calculation (they used Madyn 2000),
For the 1st critical speed, the shape mode is similar although still has difference in value. But, for the 2nd critical speed, I couldn't see it in the Ross Campbell diagram while the vendor's result has the 2nd critical speed on 2x speed mode.
Could you please help me to solve this case?

Please find below for the model that I've tried to simulate and some of vendor's analysis result:

[Model_Rotor_234.zip](https://github.com/petrobras/ross/files/15502442/Model_Rotor_234.zip)

Thank you in advance. I really appriciate your kindly response.

Regards,
Fadli

**Reply by FadliM21** [reply]
My question is why I didn't get 2nd critical speed in campbell plot diagram whereas the vendor's calculation obtained the 2nd crit speed in 2x speed mode?

Please help me to understand what's wrong with my numerical model,

Thanks in advance
Regards,
Fadli

**Reply by jguarato** [reply]
Hi @FadliM21!

It seems that the vendor's analysis only displays the torsional modes, specifically the second and third torsional modes. To obtain the torsional modes, the analysis must consider a 6 degrees of freedom (DOF) model. However, your current analysis using ROSS is based on the 4 DOF model.

To quickly convert between these models, you can use the following code:
```python
rotor_model = rotor3

shaft_elements_lenght = rotor_model.shaft_elements
shaft_elem = [
    rs.ShaftElement6DoF(
        material=rotor_model.shaft_elements[l].material,
        L=rotor_model.shaft_elements[l].L,
        n=rotor_model.shaft_elements[l].n,
        idl=rotor_model.shaft_elements[l].idl,
        odl=rotor_model.shaft_elements[l].odl,
        idr=rotor_model.shaft_elements[l].idr,
        odr=rotor_model.shaft_elements[l].odr,
        alpha=rotor_model.shaft_elements[l].alpha,
        beta=rotor_model.shaft_elements[l].beta,
        rotary_inertia=rotor_model.shaft_elements[l].rotary_inertia,
        shear_effects=rotor_model.shaft_elements[l].shear_effects,
        gyroscopic=rotor_model.shaft_elements[l].gyroscopic
    )
    for l, p in enumerate(shaft_elements_lenght)
]

disk_elem = [
    rs.DiskElement6DoF(
        n=rotor_model.disk_elements[l].n,
        m=rotor_model.disk_elements[l].m,
        Id=rotor_model.disk_elements[l].Id,
        Ip=rotor_model.disk_elements[l].Ip,
    )
    for l in range(len(rotor_model.disk_elements))
]

bearing_elem = [
    rs.BearingElement6DoF(
        n=rotor_model.bearing_elements[l].n,
        kxx=rotor_model.bearing_elements[l].kxx,
        kyy=rotor_model.bearing_elements[l].kyy,
        cxx=rotor_model.bearing_elements[l].cxx,
        cyy=rotor_model.bearing_elements[l].cyy,
        kxy=rotor_model.bearing_elements[l].kxy,
        kyx=rotor_model.bearing_elements[l].kyx,
        cxy=rotor_model.bearing_elements[l].cxy,
        cyx=rotor_model.bearing_elements[l].cyx,
        mxx=rotor_model.bearing_elements[l].mxx,
        myy=rotor_model.bearing_elements[l].myy,
        mxy=rotor_model.bearing_elements[l].mxy,
        myx=rotor_model.bearing_elements[l].myx,
        frequency=rotor_model.bearing_elements[l].frequency,
        n_link=rotor_model.bearing_elements[l].n_link,
    )
    for l in range(len(rotor_model.bearing_elements))
]

rotor6dof = rs.Rotor(shaft_elem, disk_elem, bearing_elem)
```

Please note that the identification of torsional modes in the Campbell diagram is not yet available in ROSS. We are currently working on this in PR #1066.

I ran a test with the 6 DOF model using a development version of ROSS and generated the following diagram:
```python
samples = 31
speed_range = np.linspace(0, 1900, samples)
camp = rotor6dof.run_campbell(speed_range, frequencies=21)
camp.plot(harmonics=[1, 2], frequency_units="Hz")
```
![campbel_6dof_comp_min](https://github.com/petrobras/ross/assets/82293939/a09f82fe-b251-4eb9-bbae-c8d2c51e9d52)

By hiding the other modes, a result that resembles the vendor's diagram can be observed:
![campbel_6dof_min](https://github.com/petrobras/ross/assets/82293939/e3efa975-4d1b-4fb1-9a76-a18e63852eb4)

The frequency values are not exactly the same as those shown in the vendor's results, but it seems that the model created with ROSS is slightly different from the one shown in the image you sent:
![newplot0](https://github.com/petrobras/ross/assets/82293939/f3649526-14ab-4dbb-ab62-c1e5ec01711e)

**Reply by FadliM21** [reply]
Thank you very much @jguarato , 
I'm really apreciating your helps

Regards,
Fadli

**Reply by jguarato** [reply]
You're welcome!

**Reply by FadliM21** [reply]
Hi @jguarato ,

Is the new development program of Ross (6 DoF analysis) able to be used?

Thank you

**Reply by jguarato** [reply]
Yes! To try the development version, use the following command:
```
pip install git+https://github.com/petrobras/ross.git
```

**Reply by FadliM21** [reply]
Hi @jguarato ,

How can I resize the campbell diagram to be a portrait view as your snapshot below?
![image](https://github.com/petrobras/ross/assets/165719533/57717763-bebf-477b-a6f1-1c348e012e83)
Because I always get a landscape view, so I could not create a better snapshot for the campbell diagram

Thank you in advance
Regards,

**Reply by jguarato** [reply]
You can use the _plotly_ library to export the diagram as an image by adjusting the width and height:
```python
from plotly import io as pio
fig = campbell.plot(harmonics=[1, 2], frequency_units="Hz")
pio.write_image(fig, "figure.png", width=600, height=800, scale=1)
```

**Reply by FadliM21** [reply]
Hi @jguarato ,
I have tried to use that code, unfortunately the code didn't finish the calculation, I've waited for 15 minutes and the calculation is still not done yet as the picture below:
![image](https://github.com/user-attachments/assets/849bed5d-3735-4ad6-a352-0132b286d3e0)
Could you please give me some advises for that case?

Thank you in advance

**Reply by jguarato** [reply]
I'm not sure what the problem could be in this case, as the tests I ran were successful.
![image](https://github.com/user-attachments/assets/96971b18-24a0-4c35-96ee-7fefef03e437)

Here is the link to the script I used: [[Link]](https://colab.research.google.com/drive/1IvAoOejwaLQ8DNDs2QSn1DdA18K87t4C?usp=sharing).
You may need to make a copy of my notebook and upload your ".xls" file to replicate this test.

**Reply by FadliM21** [reply]
Thank you very much @jguarato, I think there is a failed installation in my python, I'll try to reinstall again

And one more thing, I've tried to find the code to hide some of line on campbell but it is not found yet, could you please inform what the code to hide some of parameter such as forward, backward and mixed mode and remains the torsional mode?

Thank you in advance

**Reply by jguarato** [reply]
```python
for trace in fig['data']: 
    if trace['name'] in ['Forward', 'Backward', 'Mixed', 'Axial']: 
        trace['visible'] = 'legendonly'
```

**Comment by FadliM21** [comment]
Hi @jguarato 
Thank you for sharing.

Regards,
Fadli
